crates/wdk-alloc/src/lib.rs

```diff
@@ -46,32 +43,76 @@ mod kernel_mode {
+        }
+    }
 
     // SAFETY: This is safe because the Wdk allocator:
     //         1. can never unwind since it can never panic
     //         2. has implementations of alloc and dealloc that maintain layout
-    //            constraints (FIXME: Alignment of the layout is currenty not
-    //            supported)
+    //            constraints (including size and alignment)
     unsafe impl GlobalAlloc for WdkAllocator {
         unsafe fn alloc(&self, layout: Layout) -> *mut u8 {
             let ptr =
                 // SAFETY: `ExAllocatePool2` is safe to call from any `IRQL` <= `DISPATCH_LEVEL` since its allocating from `POOL_FLAG_NON_PAGED`
-                unsafe {
-                    ExAllocatePool2(POOL_FLAG_NON_PAGED, layout.size() as SIZE_T, RUST_TAG)
+                if require_realignment(layout) {
+                    // If the requested alignment is too large, we'll have to waste some space as large as the alignment.
+                    let size = layout.size() + layout.align();
+                    let mask = !(layout.align() - 1);
+                    let p = unsafe {
+                        ExAllocatePool2(POOL_FLAG_NON_PAGED, size as SIZE_T, RUST_TAG)
+                    };
+                    if !p.is_null() {
+                        let q = ((p as usize & mask) + layout.align()) as *mut PVOID;
+                        // Store the original pointer right before the pointer to return,
+                        // so that ExFreePoolWithTag can receive the correct pointer.
+                        // This is also how `_aligned_malloc` is implemented in msvcrt.dll
+                        unsafe {
+                            q.sub(1).write(p);
+                        }
+                        q.cast()
+                    } else {
+                        p
+                    }
+                } else {
+                    // Because we know the alignment while in both alloc and dealloc,
+                    // we don't need to waste space if the default alignment is small enough.
+                    unsafe {
+                        ExAllocatePool2(POOL_FLAG_NON_PAGED, layout.size() as SIZE_T, RUST_TAG)
+                    }
                 };
             if ptr.is_null() {
                 return core::ptr::null_mut();
             }
             ptr.cast()
         }
 
-        unsafe fn dealloc(&self, ptr: *mut u8, _layout: Layout) {
+        unsafe fn dealloc(&self, ptr: *mut u8, layout: Layout) {
             // SAFETY: `ExFreePool` is safe to call from any `IRQL` <= `DISPATCH_LEVEL`
             // since its freeing memory allocated from `POOL_FLAG_NON_PAGED` in `alloc`
+            let p = if require_realignment(layout) {
+                // The alignment is too large, the original pointer is stored right before
+                // the ptr. This is also how `_aligned_free` is implemented in msvcrt.dll
+                let q: *mut *mut u8 = ptr.cast();
```

Comment:
```suggestion
                let q: *mut PVOID = ptr.cast();
```

For consistency with the corresponding cast in `alloc()` (line 76)